src/lib.rs

```diff
@@ -0,0 +1,398 @@
+mod build;
+mod chrome_devtools;
+mod cmd_build;
+mod cmd_deploy;
+mod cmd_prepare_emscripten;
+mod cmd_start;
+mod cmd_test;
+mod config;
+mod deployment;
+mod emscripten;
+mod error;
+mod http_utils;
+mod package;
+mod project_dirs;
+mod test_chromium;
+mod wasm;
+mod wasm_context;
+mod wasm_export_main;
+mod wasm_export_table;
+mod wasm_gc;
+mod wasm_hook_grow;
+mod wasm_inline_js;
+mod wasm_intrinsics;
+mod wasm_js_export;
+mod wasm_js_snippet;
+mod wasm_runtime;
+
+use std::ffi::OsStr;
+use std::net::IpAddr;
+use std::path::PathBuf;
+
+use build::{Backend, BuildArgs};
+use cargo_shim::MessageFormat;
+use error::Error;
+use wasm_runtime::RuntimeKind;
+
+/// CLI for `cargo-web`
+#[derive(Debug, StructOpt)]
+#[structopt(name = "cargo-web")]
+#[structopt(about = "A `cargo` subcommand for the client-side web.")]
+#[structopt(raw(global_setting = "structopt::clap::AppSettings::ColoredHelp"))]
+#[structopt(rename_all = "kebab-case")]
+pub enum CargoWeb {
+    /// Compile a local package and all of its dependencies
+    Build {
+        #[structopt(flatten)]
+        build_args: Build,
+        #[structopt(flatten)]
+        ext: BuildExt,
+        #[structopt(flatten)]
+        build_target: Target,
+    },
+    /// Typecheck a local package and all of its dependencies
+    Check {
+        #[structopt(flatten)]
+        build_args: Build,
+        #[structopt(flatten)]
+        ext: BuildExt,
+        #[structopt(flatten)]
+        build_target: Target,
+    },
+    /// Deploys your project so that it's ready to be served statically
+    Deploy {
+        /// Output directory; the default is `$CARGO_TARGET_DIR/deploy`
+        #[structopt(short, long, parse(from_os_str))]
+        output: Option<PathBuf>,
+        #[structopt(flatten)]
+        build_args: Build,
+    },
+    /// Fetches and installs prebuilt Emscripten packages
+    PrepareEmscripten,
+    /// Runs an embedded web server, which serves the built project
+    Start {
+        /// Bind the server to this address
+        #[structopt(long, parse(try_from_str), default_value = "127.0.0.1")]
+        host: IpAddr,
+        /// Bind the server to this port
+        #[structopt(long, default_value = "8000")]
+        port: u16,
+        /// Open browser after server starts
+        #[structopt(long)]
+        open: bool,
+        /// Will try to automatically reload the page on rebuild
+        #[structopt(long)]
+        auto_reload: bool,
+        #[structopt(flatten)]
+        build_target: Target,
+        #[structopt(flatten)]
+        build_args: Build,
+    },
+    /// Compiles and runs tests
+    Test {
+        /// Compile, but don't run tests
+        #[structopt(long)]
+        no_run: bool,
+        /// Uses Node.js to run the tests
+        #[structopt(long)]
+        nodejs: bool,
+        #[structopt(flatten)]
+        build_args: Build,
+        /// all additional arguments will be passed through to the test runner
+        passthrough: Vec<String>,
+    },
+}
```

Comment:
Can we add a dummy `#[doc(hidden)]` variant to make the enum non-exhaustive? (Every publicly exported enum should be non-exhaustive so that we can safely add new entries to it in the future.)